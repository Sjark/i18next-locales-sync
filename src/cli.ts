#!/usr/bin/env node

import { Command, InvalidOptionArgumentError } from 'commander';
import path from 'path';
import fs from 'fs-extra';
import chalk from 'chalk';
import pkg from '../package.json';
import { syncLocales } from './index';

function ensureAbsolutePath(folderPath: string) {
  if (!path.isAbsolute(folderPath)) {
    folderPath = path.resolve(folderPath);
  }

  if (!fs.existsSync(folderPath)) {
    throw new InvalidOptionArgumentError(chalk.red(`\n'${folderPath}' doesn't exist`));
  }

  return folderPath;
}

const program = new Command();

program
  .version(pkg.version)
  .option('-p, --primaryLanguage <string>', 'The primary language', 'en')
  .option(
    '-s, --secondaryLanguage <string...>',
    'The secondary language, may be a list of languages'
  )
  .option(
    '-l, --localesFolder <string>',
    'The locals folder path (can be relative)',
    ensureAbsolutePath
  )
  .option(
    '-o, --outputFolder <string>',
    'The output folder, by default it will be the localesFolder',
    ensureAbsolutePath
  )
  .option('-c, --config <string>', 'The path to config file', ensureAbsolutePath);

program.parse();

const options = program.opts();

if (options.config) {
  const usersOptions = require(options.config);

  Object.assign(options, usersOptions);
}

syncLocales({
  primaryLanguage: options.primaryLanguage,
  otherLanguages: options.secondaryLanguage,
  localesFolder: options.localesFolder,
  outputFolder: options.outputFolder,
  overridePluralRules: options.overridePluralRules,
});
