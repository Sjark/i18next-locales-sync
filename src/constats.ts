export const MAX_DEPTH = 20;

export const LIB_PREFIX = '[ i18next-locales-sync ]:';
