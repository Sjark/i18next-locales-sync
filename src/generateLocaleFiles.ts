import crypto from 'crypto';
import fs from 'fs';
import * as glob from 'glob';
import path from 'path';
import { LocaleFile, LocalesFiles } from '../types/types';
import { LIB_PREFIX } from './constats';

interface Options {
  primaryLanguage: string;
  otherLanguages: string[];
  localesFolder: string;
  fileExtension: string;
}

function extractLanguagesFromPath(filepath: string, allLanguages: string[]) {
  const pathParts = filepath.split(/[\\\/]/g);
  return allLanguages.find((language) => pathParts.some((part) => part.startsWith(language)));
}

function extractNamespaceFromPath(filepath: string, language: string, fileExtension: string) {
  const filename = path.basename(filepath, fileExtension);
  if (filename === language) {
    return ''; // empty namespace
  }

  return filepath.substring(
    filepath.lastIndexOf(language) + language.length + 1,
    filepath.length - fileExtension.length
  );
}

function addMissingLanguages(localeFiles: LocalesFiles, otherLanguages: string[]) {
  otherLanguages.filter((lang) => !localeFiles[lang]).forEach((lang) => (localeFiles[lang] = {}));

  return localeFiles;
}

function addMissingNamespaces(
  localeFiles: LocalesFiles,
  primaryLanguage: string,
  otherLanguages: string[]
) {
  const primaryFiles = localeFiles[primaryLanguage];

  Object.keys(primaryFiles).forEach((namespace) => {
    otherLanguages.forEach((otherLanguage) => {
      localeFiles[otherLanguage][namespace] = localeFiles[otherLanguage][namespace] || {
        filePath: null,
        hash: '',
        data: {},
      };
    });
  });
}

export function populateFromDisk(filePath: string): LocaleFile {
  let data = {};
  let hash = '';

  if (fs.existsSync(filePath)) {
    const rawData = fs.readFileSync(filePath, 'utf8');
    data = JSON.parse(rawData);

    hash = crypto
      .createHash('md5')
      .update(JSON.stringify(data))
      .digest('hex');
  }

  return { filePath, data, hash };
}

export function generateLocaleFiles({
  localesFolder,
  fileExtension,
  primaryLanguage,
  otherLanguages,
}: Options): LocalesFiles {
  const paths = glob.sync(`${localesFolder}/**/*${fileExtension}`);
  const allLanguages = [primaryLanguage].concat(otherLanguages);

  const localeFiles = paths.reduce((structure, filePath) => {
    const remainingPath = path.relative(localesFolder, filePath);
    const language = extractLanguagesFromPath(remainingPath, allLanguages);

    if (!language) {
      return structure;
    }

    const namespace = extractNamespaceFromPath(remainingPath, language, fileExtension);

    structure[language] = structure[language] || {};
    structure[language][namespace] = populateFromDisk(filePath);

    return structure;
  }, {} as LocalesFiles);

  if (!localeFiles[primaryLanguage]) {
    throw new Error(
      `${LIB_PREFIX} There are no files for your primary language (${primaryLanguage})`
    );
  }

  addMissingLanguages(localeFiles, otherLanguages);
  addMissingNamespaces(localeFiles, primaryLanguage, otherLanguages);

  return localeFiles;
}
