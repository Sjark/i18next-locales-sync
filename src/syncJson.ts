import { JSONObject, JSONValue } from '../types/types';
import { MAX_DEPTH } from './constats';
import { isObject } from './utils/isObject';
import { traverse } from './utils/traverse';

function addMissingKeys(
  _sourceObject: JSONObject,
  targetObject: JSONObject,
  key: string,
  sourceValue: JSONValue,
  targetValue: JSONValue
) {
  if (Array.isArray(sourceValue)) {
    targetObject[key] =
      targetObject[key] && Array.isArray(targetObject[key]) ? targetObject[key] : [];
  } else if (isObject(sourceValue)) {
    targetObject[key] = targetObject[key] && isObject(targetObject[key]) ? targetObject[key] : {};
  } else {
    targetObject[key] = targetValue && !isObject(targetValue) ? targetValue : sourceValue;
  }
}

function removeKeys(targetObject: JSONObject, sourceObject: JSONObject, key: string) {
  if (sourceObject[key]) {
    // source has the target's key
    return;
  }

  if (Array.isArray(targetObject)) {
    targetObject.splice(+key, 1);
  } else if (isObject(targetObject)) {
    delete targetObject[key];
  }
}

export function syncJson(source: JSONObject, target: JSONObject, depth = MAX_DEPTH) {
  traverse(source, target, addMissingKeys, depth);
  traverse(target, source, removeKeys, depth);

  return target;
}
