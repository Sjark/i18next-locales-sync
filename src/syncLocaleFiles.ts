import { LocalesFiles } from '../types/types';
import { syncJson } from './syncJson';

interface Options {
  localeFiles: LocalesFiles;
  primaryLanguage: string;
  otherLanguages: string[];
}

export function syncLocaleFiles({
  localeFiles,
  primaryLanguage,
  otherLanguages,
}: Options): LocalesFiles {
  const primaryLocaleFiles = localeFiles[primaryLanguage];
  otherLanguages.forEach((currentLanguage) => {
    const currentNamespaces = localeFiles[currentLanguage];

    Object.keys(primaryLocaleFiles).forEach((primaryNamespace) => {
      const { data } = syncJson({
        source: { data: primaryLocaleFiles[primaryNamespace].data, language: primaryLanguage },
        target: { data: currentNamespaces[primaryNamespace].data, language: currentLanguage },
      });

      currentNamespaces[primaryNamespace].data = data;
    });
  });

  return localeFiles;
}
