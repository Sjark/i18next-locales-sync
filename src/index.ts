import { LIB_PREFIX } from './constats';
import { generateLocaleFiles } from './generateLocaleFiles';
import { PluralResolver } from './i18next/PluralResolver';
import { syncLocaleFiles } from './syncLocaleFiles';
import { writeToDisk } from './utils/writeToDisk';
import chalk from 'chalk';

interface SyncLocalesOptions {
  primaryLanguage: string;
  otherLanguages: string[];
  localesFolder: string;
  outputFolder?: string;
  fileExtension?: string;
  pluralResolver?: PluralResolver;
}

export function syncLocales({
  primaryLanguage,
  otherLanguages,
  localesFolder,
  outputFolder = localesFolder,
  fileExtension = '.json',
  pluralResolver,
}: SyncLocalesOptions) {
  const localeFiles = generateLocaleFiles({
    primaryLanguage,
    otherLanguages,
    localesFolder,
    fileExtension,
  });

  syncLocaleFiles({ localeFiles, primaryLanguage, otherLanguages, pluralResolver });

  writeToDisk({ localeFiles, primaryLanguage, otherLanguages, outputFolder, localesFolder });

  console.log(
    chalk.green`${chalk.bold.greenBright(LIB_PREFIX)} '${localesFolder}' was synced successfully.`
  );
}
