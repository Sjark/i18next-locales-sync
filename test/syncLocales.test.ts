import path from 'path';
import { syncLocales } from '../src';
import { vol } from 'memfs';

describe('syncLocales - E2E', () => {
  it('should sync locale files without namespaces', () => {
    const primaryLanguage = 'en';
    const otherLanguages = ['ja', 'he', 'de'];
    const outputFolder = path.resolve('./test/output/fixture1');

    syncLocales({
      primaryLanguage,
      otherLanguages,
      localesFolder: path.resolve('./test/fixtures/fixture1'),
      outputFolder,
      fileExtension: '.json',
    });

    expect(vol.toJSON(outputFolder)).toMatchSnapshot();
  });

  it('should sync locale files with namespaces', () => {
    const primaryLanguage = 'en';
    const otherLanguages = ['ja', 'he', 'de'];
    const outputFolder = path.resolve('./test/output/fixture2');

    syncLocales({
      primaryLanguage,
      otherLanguages,
      localesFolder: path.resolve('./test/fixtures/fixture2'),
      outputFolder,
      fileExtension: '.json',
    });

    expect(vol.toJSON(outputFolder)).toMatchSnapshot();
  });

  it('should sync locale files with nested namespaces', () => {
    const primaryLanguage = 'en';
    const otherLanguages = ['ja', 'he', 'de'];
    const outputFolder = path.resolve('./test/output/fixture3');

    syncLocales({
      primaryLanguage,
      otherLanguages,
      localesFolder: path.resolve('./test/fixtures/fixture3'),
      outputFolder,
      fileExtension: '.json',
    });

    expect(vol.toJSON(outputFolder)).toMatchSnapshot();
  });
});
