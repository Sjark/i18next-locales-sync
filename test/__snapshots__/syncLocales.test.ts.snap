// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syncLocales - E2E should sync locale files with different folder structure 1`] = `
Object {
  "test/output/fixture4/common/de.json": "{\\"test\\":\\"bla-en\\",\\"test_plural\\":\\"bla-plural\\"}",
  "test/output/fixture4/common/en.json": "{\\"test\\":\\"bla-en\\",\\"test_plural\\":\\"bla-plural\\"}",
  "test/output/fixture4/common/he.json": "{\\"test\\":\\"bla-he\\",\\"test_1\\":\\"bla-1\\",\\"test_0\\":\\"bla-plural\\",\\"test_2\\":\\"bla-plural\\",\\"test_3\\":\\"bla-plural\\"}",
  "test/output/fixture4/common/ja.json": "{\\"test\\":\\"bla-en\\"}",
  "test/output/fixture4/front/de.json": "{\\"test\\":\\"bla-he\\"}",
  "test/output/fixture4/front/en.json": "{\\"test\\":\\"bla-he\\"}",
  "test/output/fixture4/front/he.json": "{\\"test\\":\\"bla\\",\\"test_0\\":\\"bla-0\\"}",
  "test/output/fixture4/front/ja.json": "{\\"test\\":\\"bla-he\\"}",
}
`;

exports[`syncLocales - E2E should sync locale files with namespaces 1`] = `
Object {
  "/Users/catz/Projects/oss/i18next-locales-sync/test/output/fixture2/de/common.json": "{\\"test\\":\\"bla-de\\",\\"test_plural\\":\\"bla-plural\\"}",
  "/Users/catz/Projects/oss/i18next-locales-sync/test/output/fixture2/de/front.json": "{\\"title\\":\\"front\\"}",
  "/Users/catz/Projects/oss/i18next-locales-sync/test/output/fixture2/en/common.json": "{\\"test\\":\\"bla\\",\\"test_plural\\":\\"bla-plural\\"}",
  "/Users/catz/Projects/oss/i18next-locales-sync/test/output/fixture2/en/front.json": "{\\"title\\":\\"front\\"}",
  "/Users/catz/Projects/oss/i18next-locales-sync/test/output/fixture2/he/common.json": "{\\"test\\":\\"bla\\",\\"test_0\\":\\"bla-plural\\",\\"test_1\\":\\"bla-plural\\",\\"test_2\\":\\"bla-plural\\",\\"test_3\\":\\"bla-plural\\"}",
  "/Users/catz/Projects/oss/i18next-locales-sync/test/output/fixture2/he/front.json": "{\\"title\\":\\"front\\"}",
  "/Users/catz/Projects/oss/i18next-locales-sync/test/output/fixture2/ja/common.json": "{\\"test\\":\\"bla\\"}",
  "/Users/catz/Projects/oss/i18next-locales-sync/test/output/fixture2/ja/front.json": "{\\"title\\":\\"front\\"}",
}
`;

exports[`syncLocales - E2E should sync locale files with nested namespaces 1`] = `
Object {
  "test/output/fixture3/de/common.json": "{\\"test\\":\\"bla-de\\",\\"test_plural\\":\\"bla-plural\\"}",
  "test/output/fixture3/de/nested/a.json": "{\\"a\\":\\"bla-de\\"}",
  "test/output/fixture3/de/nested/b.json": "{\\"b\\":\\"bla-en\\"}",
  "test/output/fixture3/en/common.json": "{\\"test\\":\\"bla\\",\\"test_plural\\":\\"bla-plural\\"}",
  "test/output/fixture3/en/nested/a.json": "{\\"a\\":\\"bla-en\\"}",
  "test/output/fixture3/en/nested/b.json": "{\\"b\\":\\"bla-en\\"}",
  "test/output/fixture3/he/common.json": "{\\"test\\":\\"bla-he\\",\\"test_1\\":\\"bla-plural-he\\",\\"test_0\\":\\"bla-plural\\",\\"test_2\\":\\"bla-plural\\",\\"test_3\\":\\"bla-plural\\"}",
  "test/output/fixture3/he/nested/a.json": "{\\"a\\":\\"bla-en\\"}",
  "test/output/fixture3/he/nested/b.json": "{\\"b\\":\\"bla-en\\"}",
  "test/output/fixture3/ja/common.json": "{\\"test\\":\\"bla\\"}",
  "test/output/fixture3/ja/nested/a.json": "{\\"a\\":\\"bla-en\\"}",
  "test/output/fixture3/ja/nested/b.json": "{\\"b\\":\\"bla-en\\"}",
}
`;

exports[`syncLocales - E2E should sync locale files without namespaces 1`] = `
Object {
  "test/output/fixture1/de.json": "{\\"test\\":\\"bla\\",\\"test_plural\\":\\"bla-plural\\"}",
  "test/output/fixture1/en.json": "{\\"test\\":\\"bla\\",\\"test_plural\\":\\"bla-plural\\"}",
  "test/output/fixture1/he.json": "{\\"test\\":\\"bla\\",\\"test_0\\":\\"bla-0\\",\\"test_1\\":\\"bla-plural\\",\\"test_2\\":\\"bla-plural\\",\\"test_3\\":\\"bla-plural\\"}",
  "test/output/fixture1/ja.json": "{\\"test\\":\\"bla\\"}",
}
`;

exports[`syncLocales - E2E should sync locales to the same locales folder 1`] = `
Object {
  "test/fixtures/fixture3/de/common.json": "{\\"test\\":\\"bla-de\\",\\"test_plural\\":\\"bla-plural\\"}",
  "test/fixtures/fixture3/de/nested/a.json": "{
  \\"a\\": \\"bla-de\\"
}
",
  "test/fixtures/fixture3/de/nested/b.json": "{\\"b\\":\\"bla-en\\"}",
  "test/fixtures/fixture3/en/common.json": "{
  \\"test\\": \\"bla\\",
  \\"test_plural\\": \\"bla-plural\\"
}
",
  "test/fixtures/fixture3/en/nested/a.json": "{
  \\"a\\": \\"bla-en\\"
}
",
  "test/fixtures/fixture3/en/nested/b.json": "{
  \\"b\\": \\"bla-en\\"
}
",
  "test/fixtures/fixture3/he/common.json": "{\\"test\\":\\"bla-he\\",\\"test_1\\":\\"bla-plural-he\\",\\"test_0\\":\\"bla-plural\\",\\"test_2\\":\\"bla-plural\\",\\"test_3\\":\\"bla-plural\\"}",
  "test/fixtures/fixture3/he/nested/a.json": "{\\"a\\":\\"bla-en\\"}",
  "test/fixtures/fixture3/he/nested/b.json": "{\\"b\\":\\"bla-en\\"}",
  "test/fixtures/fixture3/ja/common.json": "{\\"test\\":\\"bla\\"}",
  "test/fixtures/fixture3/ja/nested/a.json": "{\\"a\\":\\"bla-en\\"}",
  "test/fixtures/fixture3/ja/nested/b.json": "{\\"b\\":\\"bla-en\\"}",
}
`;
